### PR TITLE
Permit shared region as key or value for map

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #620

This pull request introduces changes in the `src/assertions.cpp`, `src/crab/ebpf_domain.cpp`, `test-data/call.yaml`, and `test-data/callx.yaml` files. The changes mainly focus on modifying the handling of memory types in the codebase and updating the test cases accordingly. 

Changes to memory handling:

* [`src/assertions.cpp`](diffhunk://#diff-6ce2751805a1c8beb6524a37731195a873e763cbeaf8110ba4f60645d65b6facL73-R73): Modified the `AssertExtractor` class to change the type constraint of `arg.reg` from `TypeGroup::stack_or_packet` to `TypeGroup::mem` for `PTR_TO_MAP_KEY` and `PTR_TO_MAP_VALUE` cases.
* [`src/crab/ebpf_domain.cpp`](diffhunk://#diff-4260cca82870d77e078c224955ebce890cb127bdd92fd63a2d57e1374dedd9f2R1756-R1761): Updated the `ebpf_domain_t::operator()` method to handle `T_SHARED` access register type. It now checks the access to shared memory and requires that the shared value is greater than 0.

Updates to test cases:

* [`test-data/call.yaml`](diffhunk://#diff-f5adec440910e830f3e5bbdaf3336d42877c068c3271a74ac4a306fa901a3f93R43-R133): Added new test cases for the `bpf_map_update_elem` function with different conditions for shared values. Also, updated the error messages in existing test cases to include `shared` as a valid type. [[1]](diffhunk://#diff-f5adec440910e830f3e5bbdaf3336d42877c068c3271a74ac4a306fa901a3f93R43-R133) [[2]](diffhunk://#diff-f5adec440910e830f3e5bbdaf3336d42877c068c3271a74ac4a306fa901a3f93L79-R170)
* [`test-data/callx.yaml`](diffhunk://#diff-a4a0d84cbb86eab541265a66885b8fab6424032749d1b574f242763dc6448d2fL143-R143): Updated the error message in a test case to include `shared` as a valid type.

### CODE REVIEW

File: test-data/call.yaml [hunk 1]
```diff
@@ -40,6 +40,97 @@ post:
...
+code:
+  <start>: |
+    call 2; bpf_map_update_elem
+
+post:
+  - r0.type=number
+  - s[496...511].type=number
+
+---
+test-case: bpf_map_update_elem with shared value
```
Comment (dthaler) [inline test-data/call.yaml:60]: what about a shared _key_?

File: test-data/callx.yaml [hunk 2]
```diff
@@ -140,7 +140,7 @@ post:
   - r8.uvalue=2
 
 messages:
-  - "0: Invalid type (r3.type in {stack, packet})"
+  - "0: Invalid type (r3.type in {stack, packet, shared})"
```
Comment (dthaler) [inline test-data/callx.yaml:143]: also copy the new test cases into this file, e.g., "bpf_map_update_elem with numeric stack value"

File: test-data/call.yaml (hunk 1 above)
Comment (Alan-Jowett) [reply test-data/call.yaml:158]: Added shared key and packet cases.

File: test-data/callx.yaml (hunk 2 above)
Comment (Alan-Jowett) [reply test-data/callx.yaml:394]: Done.

**Review by dthaler — APPROVED** [review]

**Pull request merged**